### PR TITLE
feat(frontend): Pushdown condition above join to other side if all `InputRef`s are eq-condition columns

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Pushdown join condition to other side if all `InputRef`s are eq-condition columns

Questions:
1. How to handle `null_safe` flag: we can apply the condition on nullsafe condition (NOT DISTINCT FROM) since the condition pushed down will filter out/not filter out rows with equal keys on both sides. This applies even if the condition is pushed from outer side to inner side. The only way that `IS NOT DISTINCT FROM` differs from `=` is that the former allows `NULL = NULL`, so we only need to care about how NULL is handled on both sides.
2. How to handle non-inner joins: we can only handle pushing down to an inner side. If the side is outer, we do not want to apply filter to it, since we do not want to filter a row which will be forwarded regardless of match on other side. Conversely, if filter originally from outer side filters row on outer side, then we will never forward a corresponding row on inner side, since we require match with outer side to forward.

TODO:
- X Unit test
-  [X] More planner test (test more types of predicates, test non-inner joins?)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
Fix: https://github.com/risingwavelabs/risingwave/issues/6862

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/subquery_expr_correlated.yaml [hunk 1]
```diff
@@ -461,11 +461,11 @@
           └─LogicalAgg { group_key: [a.a3, b.b2], aggs: [count(1:Int32)] }
             └─LogicalJoin { type: LeftOuter, on: IsNotDistinctFrom(a.a3, c.c3) AND IsNotDistinctFrom(b.b2, c.c2), output: [a.a3, b.b2, 1:Int32] }
               ├─LogicalAgg { group_key: [a.a3, b.b2], aggs: [] }
-              | └─LogicalJoin { type: Inner, on: true, output: all }
+              | └─LogicalJoin { type: Inner, on: (a.a3 = b.b2), output: all }
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/subquery_expr_correlated.yaml:464]: Not supposed to happen. Lets investigate why (might be to do with subquery).

**Comment by jon-chuang** [comment]
Seems this PR also caused during deterministic-simulation/fuzzing: 
```
thread '<unnamed>' panicked at 'hash key should not be empty, use `Single` instead'
```

File: src/frontend/planner_test/tests/testdata/subquery_expr_correlated.yaml (hunk 1 above)
Comment (jon-chuang) [reply src/frontend/planner_test/tests/testdata/subquery_expr_correlated.yaml:464]: Actually, this is correct. The filter was actually derived from the inner join above. If you stare at the plan long enough you will see that the filter derivation and pushdown is indeed correct:
- we want to filter out those rows of `a, b` such that a3 = b2, and that table c has 3 elements such that `c2 = b2 = a3 = c3`

**Review by st1page — COMMENTED** [review]
The PR LGTM. The optimization seems a subset of [predicate move around](https://www.vldb.org/conf/1994/P096.PDF). So do we need to implement it completely :thinking: @chenzl25?

**Comment by chenzl25** [comment]
> The PR LGTM. The optimization seems a subset of [predicate move around](https://www.vldb.org/conf/1994/P096.PDF). So do we need to implement it completely 🤔 @chenzl25?

Yes. Currently we can only push the predicate in a top-down way. In order to make it possible to move the predicates from one side of the join to the other side, I think we should implement predicate pulled up as well. We can use an optimizer rule to match a `LogicalJoin` and try to pull up its predicates and then distribute the inferred predicates to the join inputs.

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 2]
```diff
@@ -842,6 +912,56 @@ impl PredicatePushdown for LogicalJoin {
...
+        // Derive conditions to push to the other side based on eq condition columns
+        let left_cols_num = self.left().schema().len();
+        let right_cols_num = self.right().schema().len();
+        let eq_condition = EqJoinPredicate::create(left_cols_num, right_cols_num, new_on.clone());
+
+        // Only push to RHS if RHS is inner side of a join (RHS requires match on LHS)
+        // TODO: make this a function
+        let right_derived_predicate = if matches!(
+            join_type,
+            JoinType::Inner | JoinType::LeftOuter | JoinType::RightSemi
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:924]: It seems `LeftSemiJoin` and `RightSemiJoin` are both acceptable here.

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 3]
```diff
@@ -842,6 +912,56 @@ impl PredicatePushdown for LogicalJoin {
...
+                    .collect(),
+            }
+        } else {
+            Condition::true_cond()
+        };
+
+        // Only push to LHS if LHS is inner side of a join (LHS requires match on RHS)
+        let left_derived_predicate = if matches!(
+            join_type,
+            JoinType::Inner | JoinType::RightOuter | JoinType::LeftSemi
```
Comment (chenzl25) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:942]: Ditto.

**Comment by jon-chuang** [comment]
> Currently we can only push the predicate in a top-down way. In order to make it possible to move the predicates from one side of the join to the other side, I think we should implement predicate pulled up as well.

I believe we already do this for multiway inner joins. Not sure how applicable predicate pullup is for non-inner joins.

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 2 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:924]: For the `LeftSemi` case, indeed, we will never need a row on RHS if matching rows on LHS are filtered out.

**Comment by chenzl25** [comment]
> > Currently we can only push the predicate in a top-down way. In order to make it possible to move the predicates from one side of the join to the other side, I think we should implement predicate pulled up as well.
> 
> I believe we already do this for multiway inner joins. Not sure how applicable predicate pullup is for non-inner joins.
> 
> The paper only mentions predicate pullup for: GroupBy, Select (Project) and Union/Intersection nodes.

For outer join, we can pull up the predicates from outer side and then push them down to the inner side (Both sides of full outer join are outer). All the predicate pullup logical just need to be sound but not complete.

**Comment by jon-chuang** [comment]
Hello, I decided to define pure function as limited to a certain enum range of the proto. Secondly, I added planner tests. PTAL, thanks!

**Comment by lmatz** [comment]
`license-header-check ` is not working and blocks many PRs to be merged after this one, 
let's try to merge the next one first.

**Pull request merged**